File: crates/client/src/pairer.rs
use media_elo_core::Row;
use rand::seq::SliceRandom;
use rand::Rng;
use std::collections::VecDeque;
use uuid::Uuid;

pub const HISTORY_SIZE: usize = 30;
pub const RANDOM_PAIR_CHANCE: f64 = 0.2;
pub const PAIR_ATTEMPTS: usize = 20;

pub struct Pairer {
    recent: VecDeque<(Uuid, Uuid)>,
    history_size: usize,
}

impl Default for Pairer {
    fn default() -> Self {
        Self::new()
    }
}

impl Pairer {
    pub fn new() -> Self {
        Self {
            recent: VecDeque::with_capacity(HISTORY_SIZE),
            history_size: HISTORY_SIZE,
        }
    }

    fn pair_id(a: &Row, b: &Row) -> (Uuid, Uuid) {
        if a.id <= b.id {
            (a.id, b.id)
        } else {
            (b.id, a.id)
        }
    }

    fn pick_candidate_idx<R: Rng>(rng: &mut R, rows: &[Row], pool: &[usize]) -> usize {
        let weights: Vec<f64> = pool
            .iter()
            .map(|&i| 1.0 / (1.0 + rows[i].matches as f64))
            .collect();
        weighted_pick_idx(rng, pool, &weights)
    }

    fn weighted_opponent_idx<R: Rng>(
        rng: &mut R,
        rows: &[Row],
        a_idx: usize,
        candidates: &[usize],
    ) -> usize {
        let a_elo = rows[a_idx].elo;
        let weights: Vec<f64> = candidates
            .iter()
            .map(|&i| {
                let b = &rows[i];
                (1.0 / (1.0 + b.matches as f64))
                    * (1.0 / (1.0 + (a_elo - b.elo).abs()))
            })
            .collect();
        weighted_pick_idx(rng, candidates, &weights)
    }

    /// Picks a pair from `eligible` (indices into `rows`). Returned values are row IDs.
    pub fn pick(&self, rows: &[Row], eligible: &[usize]) -> Option<(Uuid, Uuid)> {
        if eligible.len() < 2 {
            return None;
        }
        let mut rng = rand::thread_rng();

        let a_idx = Self::pick_candidate_idx(&mut rng, rows, eligible);
        let a_type = rows[a_idx].type_.clone();

        let same_type: Vec<usize> = eligible
            .iter()
            .copied()
            .filter(|&i| i != a_idx && rows[i].type_ == a_type)
            .collect();
        if same_type.is_empty() {
            return None;
        }

        let unmatched: Vec<usize> = same_type
            .iter()
            .copied()
            .filter(|&i| rows[i].matches == 0)
            .collect();

        if !unmatched.is_empty() {
            for _ in 0..PAIR_ATTEMPTS {
                let &b_idx = unmatched.choose(&mut rng).unwrap();
                let pid = Self::pair_id(&rows[a_idx], &rows[b_idx]);
                if !self.recent.contains(&pid) {
                    return Some((rows[a_idx].id, rows[b_idx].id));
                }
            }
        }

        for _ in 0..PAIR_ATTEMPTS {
            let b_idx = if rng.gen::<f64>() < RANDOM_PAIR_CHANCE {
                *same_type.choose(&mut rng).unwrap()
            } else {
                Self::weighted_opponent_idx(&mut rng, rows, a_idx, &same_type)
            };
            let pid = Self::pair_id(&rows[a_idx], &rows[b_idx]);
            if !self.recent.contains(&pid) {
                return Some((rows[a_idx].id, rows[b_idx].id));
            }
        }
        None
    }

    pub fn forget_last(&mut self) -> Option<(Uuid, Uuid)> {
        self.recent.pop_back()
    }

    pub fn remember(&mut self, a: &Row, b: &Row) {
        if self.recent.len() == self.history_size {
            self.recent.pop_front();
        }
        self.recent.push_back(Self::pair_id(a, b));
    }
}

fn weighted_pick_idx<R: Rng>(rng: &mut R, pool: &[usize], weights: &[f64]) -> usize {
    let total: f64 = weights.iter().sum();
    if total <= 0.0 {
        return pool[0];
    }
    let mut t = rng.gen::<f64>() * total;
    for (i, w) in weights.iter().enumerate() {
        t -= w;
        if t <= 0.0 {
            return pool[i];
        }
    }
    pool[pool.len() - 1]
}

File: android/app/src/main/java/com/mediaelo/client/data/Pairer.kt
package com.mediaelo.client.data

import com.mediaelo.client.api.Row
import kotlin.math.abs
import kotlin.random.Random

/**
 * Port of rust/crates/client/src/pairer.rs.
 *
 * Strategy:
 *  - Pick a candidate A weighted by 1/(1+matches) so under-rated rows surface.
 *  - Pick B from same type, preferring unmatched first, then weighted by
 *    1/(1+matches) × 1/(1+|elo_a - elo_b|) with a 20% pure-random fallback.
 *  - Avoid the last 30 pairs by ID to keep things fresh.
 */
class Pairer(
    private val historySize: Int = 30,
    private val pairAttempts: Int = 20,
    private val randomPairChance: Double = 0.2,
    private val rng: Random = Random.Default,
) {
    private val recent: ArrayDeque<Pair<String, String>> = ArrayDeque(historySize)

    private fun pairId(a: Row, b: Row): Pair<String, String> =
        if (a.id <= b.id) a.id to b.id else b.id to a.id

    fun pick(rows: List<Row>, eligibleIdx: List<Int>): Pair<Row, Row>? {
        if (eligibleIdx.size < 2) return null

        val aIdx = pickCandidateIdx(rows, eligibleIdx)
        val aType = rows[aIdx].type

        val sameType = eligibleIdx.filter { it != aIdx && rows[it].type == aType }
        if (sameType.isEmpty()) return null

        val unmatched = sameType.filter { rows[it].matches == 0 }
        if (unmatched.isNotEmpty()) {
            repeat(pairAttempts) {
                val bIdx = unmatched.random(rng)
                val pid = pairId(rows[aIdx], rows[bIdx])
                if (pid !in recent) return rows[aIdx] to rows[bIdx]
            }
        }

        repeat(pairAttempts) {
            val bIdx = if (rng.nextDouble() < randomPairChance) {
                sameType.random(rng)
            } else {
                weightedOpponentIdx(rows, aIdx, sameType)
            }
            val pid = pairId(rows[aIdx], rows[bIdx])
            if (pid !in recent) return rows[aIdx] to rows[bIdx]
        }
        return null
    }

    fun remember(a: Row, b: Row) {
        if (recent.size == historySize) recent.removeFirst()
        recent.addLast(pairId(a, b))
    }

    fun forgetLast() { recent.removeLastOrNull() }

    private fun pickCandidateIdx(rows: List<Row>, pool: List<Int>): Int {
        val weights = pool.map { 1.0 / (1.0 + rows[it].matches) }
        return weightedPickIdx(pool, weights)
    }

    private fun weightedOpponentIdx(rows: List<Row>, aIdx: Int, candidates: List<Int>): Int {
        val aElo = rows[aIdx].elo
        val weights = candidates.map { i ->
            val b = rows[i]
            (1.0 / (1.0 + b.matches)) * (1.0 / (1.0 + abs(aElo - b.elo)))
        }
        return weightedPickIdx(candidates, weights)
    }

    private fun weightedPickIdx(pool: List<Int>, weights: List<Double>): Int {
        val total = weights.sum()
        if (total <= 0.0) return pool.first()
        var t = rng.nextDouble() * total
        for (i in weights.indices) {
            t -= weights[i]
            if (t <= 0.0) return pool[i]
        }
        return pool.last()
    }
}

private val RANKABLE = setOf("done", "dropped")

fun List<Row>.eligibleIndices(typeFilter: String? = null): List<Int> =
    withIndex()
        .filter { (_, r) -> r.status in RANKABLE && (typeFilter == null || r.type == typeFilter) }
        .map { it.index }
